File: packages/plugin-swc/src/internal/ast/metadata.ts
/**
 * Metadata collection for gql definitions in SWC AST.
 *
 * Traverses the SWC AST and collects metadata about gql.default() calls,
 * including their canonical paths, export status, and depth in the AST.
 */

import type { CanonicalPathTracker } from "@soda-gql/builder";
import { createCanonicalTracker } from "@soda-gql/builder";
import type { GqlDefinitionMetadata } from "@soda-gql/plugin-common";
import type {
  ArrowFunctionExpression,
  AssignmentExpression,
  CallExpression,
  ClassDeclaration,
  Expression,
  FunctionDeclaration,
  FunctionExpression,
  MemberExpression,
  Module,
  ModuleItem,
  VariableDeclarator,
} from "@swc/types";

export type GqlDefinitionMetadataMap = WeakMap<CallExpression, GqlDefinitionMetadata>;

type CanonicalTrackerFactory = typeof createCanonicalTracker;

type CollectArgs = {
  readonly module: Module;
  readonly filename: string;
  readonly createTracker?: CanonicalTrackerFactory;
};

type ScopeHandle = ReturnType<CanonicalPathTracker["enterScope"]>;
type ExportBindingMap = Map<string, string>;

export const collectGqlDefinitionMetadata = ({ module, filename, createTracker }: CollectArgs): GqlDefinitionMetadataMap => {
  const exportBindings = collectExportBindings(module.body);
  const trackerFactory = createTracker ?? createCanonicalTracker;
  const tracker = trackerFactory({
    filePath: filename,
    getExportName: (localName) => exportBindings.get(localName),
  });

  const getAnonymousName = createAnonymousNameFactory();
  const scopeStack: ScopeHandle[] = [];
  const metadata = new WeakMap<CallExpression, GqlDefinitionMetadata>();

  // Traverse the module
  const traverse = (node: unknown): void => {
    if (!node || typeof node !== "object") {
      return;
    }

    // Handle CallExpression nodes - check if it's a gql definition
    if (isCallExpression(node) && isGqlDefinitionCall(node)) {
      const depthBeforeRegister = tracker.currentDepth();
      const { astPath } = tracker.registerDefinition();
      const isTopLevel = depthBeforeRegister <= 1;
      const exportInfo = isTopLevel ? resolveTopLevelExport(node, exportBindings, module.body) : null;

      metadata.set(node, {
        astPath,
        isTopLevel,
        isExported: exportInfo?.isExported ?? false,
        exportBinding: exportInfo?.exportBinding,
      });

      // Skip traversing the body of gql definitions
      return;
    }

    // Enter scope if applicable
    const handle = maybeEnterScope(node, tracker, getAnonymousName);
    if (handle) {
      scopeStack.push(handle);
    }

    // Recursively traverse children
    if (Array.isArray(node)) {
      for (const child of node) {
        traverse(child);
      }
    } else {
      for (const value of Object.values(node)) {
        if (value && typeof value === "object") {
          traverse(value);
        }
      }
    }

    // Exit scope if we entered one
    if (handle) {
      const poppedHandle = scopeStack.pop();
      if (poppedHandle) {
        tracker.exitScope(poppedHandle);
      }
    }
  };

  traverse(module);

  return metadata;
};

/**
 * Collect export bindings from the module.
 */
const collectExportBindings = (body: ModuleItem[]): ExportBindingMap => {
  const bindings: ExportBindingMap = new Map();

  for (const statement of body) {
    // ESM exports: export const foo = ...
    if (statement.type === "ExportDeclaration" && statement.declaration) {
      const decl = statement.declaration;
      if (decl.type === "VariableDeclaration") {
        for (const declarator of decl.declarations) {
          if (declarator.id.type === "Identifier") {
            bindings.set(declarator.id.value, declarator.id.value);
          }
        }
        continue;
      }

      if (
        (decl.type === "FunctionDeclaration" || decl.type === "ClassDeclaration") &&
        decl.identifier &&
        decl.identifier.type === "Identifier"
      ) {
        bindings.set(decl.identifier.value, decl.identifier.value);
      }
      continue;
    }

    // CommonJS exports: exports.foo = ... or module.exports.foo = ...
    if (statement.type === "ExpressionStatement" && statement.expression.type === "AssignmentExpression") {
      const exportName = getCommonJsExportName(statement.expression.left);
      if (exportName) {
        bindings.set(exportName, exportName);
      }
    }
  }

  return bindings;
};

/**
 * Extract CommonJS export name from assignment left-hand side.
 */
const getCommonJsExportName = (node: unknown): string | null => {
  if (typeof node !== "object" || node === null || !("type" in node) || node.type !== "MemberExpression") {
    return null;
  }

  const memberExpr = node as MemberExpression;

  if (memberExpr.property.type === "Computed") {
    return null;
  }

  // Check if it's exports.foo or module.exports.foo
  const isExports = memberExpr.object.type === "Identifier" && memberExpr.object.value === "exports";
  const isModuleExports =
    memberExpr.object.type === "MemberExpression" &&
    memberExpr.object.object.type === "Identifier" &&
    memberExpr.object.object.value === "module" &&
    memberExpr.object.property.type === "Identifier" &&
    memberExpr.object.property.value === "exports";

  if (!isExports && !isModuleExports) {
    return null;
  }

  // Extract property name
  if (memberExpr.property.type === "Identifier") {
    return memberExpr.property.value;
  }

  return null;
};

/**
 * Create factory for generating anonymous names.
 */
const createAnonymousNameFactory = (): ((kind: string) => string) => {
  const counters = new Map<string, number>();
  return (kind) => {
    const count = counters.get(kind) ?? 0;
    counters.set(kind, count + 1);
    return `${kind}#${count}`;
  };
};

/**
 * Check if a node is a gql definition call.
 */
const isGqlDefinitionCall = (node: CallExpression): boolean => {
  if (node.callee.type !== "MemberExpression") {
    return false;
  }
  if (!isExpression(node.callee.object)) {
    return false;
  }
  if (!isGqlReference(node.callee.object)) {
    return false;
  }
  if (node.arguments.length === 0) {
    return false;
  }
  const firstArg = node.arguments[0];
  if (!firstArg) {
    return false;
  }
  return firstArg.expression.type === "ArrowFunctionExpression";
};

/**
 * Check if an expression is a reference to `gql`.
 */
const isGqlReference = (expr: Expression): boolean => {
  if (expr.type === "Identifier" && expr.value === "gql") {
    return true;
  }
  if (expr.type !== "MemberExpression") {
    return false;
  }
  if (expr.property.type === "Computed") {
    return false;
  }
  if (expr.property.type === "Identifier" && expr.property.value === "gql") {
    return true;
  }
  // Check if object is also an Expression before recursing
  if (!isExpression(expr.object)) {
    return false;
  }
  return isGqlReference(expr.object);
};

/**
 * Resolve top-level export information for a call expression.
 */
const resolveTopLevelExport = (
  callExpr: CallExpression,
  exportBindings: ExportBindingMap,
  body: ModuleItem[],
): { readonly isExported: true; readonly exportBinding: string } | null => {
  // Find the parent statement/declaration
  for (const item of body) {
    // ESM: export const foo = gql.default(...)
    if (item.type === "ExportDeclaration" && item.declaration?.type === "VariableDeclaration") {
      for (const declarator of item.declaration.declarations) {
        const init = declarator.init;
        if (init && containsCallExpression(init, callExpr)) {
          if (declarator.id.type === "Identifier") {
            return { isExported: true, exportBinding: declarator.id.value };
          }
        }
      }
    }

    // ESM: const foo = gql.default(...); export { foo }
    if (item.type === "VariableDeclaration") {
      for (const declarator of item.declarations) {
        const init = declarator.init;
        if (init && containsCallExpression(init, callExpr)) {
          if (declarator.id.type === "Identifier") {
            const exportBinding = exportBindings.get(declarator.id.value);
            if (exportBinding) {
              return { isExported: true, exportBinding };
            }
          }
        }
      }
    }

    // CommonJS: exports.foo = gql.default(...)
    if (item.type === "ExpressionStatement" && item.expression.type === "AssignmentExpression") {
      const right = item.expression.right;
      if (containsCallExpression(right, callExpr)) {
        const exportName = getCommonJsExportName(item.expression.left);
        if (exportName && exportBindings.has(exportName)) {
          return { isExported: true, exportBinding: exportName };
        }
      }
    }
  }

  return null;
};

/**
 * Check if an expression contains a specific call expression.
 */
const containsCallExpression = (expr: unknown, target: CallExpression): boolean => {
  if (expr === target) {
    return true;
  }
  if (typeof expr !== "object" || !expr) {
    return false;
  }
  for (const value of Object.values(expr)) {
    if (value && typeof value === "object") {
      if (containsCallExpression(value, target)) {
        return true;
      }
    }
  }
  return false;
};

/**
 * Enter a scope if applicable for the given node.
 */
const maybeEnterScope = (
  node: unknown,
  tracker: CanonicalPathTracker,
  getAnonymousName: (kind: string) => string,
): ScopeHandle | null => {
  if (!node || typeof node !== "object") {
    return null;
  }

  // CommonJS exports: exports.foo = ... or module.exports.foo = ...
  if (isAssignmentExpression(node)) {
    const exportName = getCommonJsExportName(node.left);
    if (exportName) {
      return tracker.enterScope({ segment: exportName, kind: "variable", stableKey: `var:${exportName}` });
    }
  }

  if (isVariableDeclarator(node) && node.id.type === "Identifier") {
    const name = node.id.value;
    return tracker.enterScope({ segment: name, kind: "variable", stableKey: `var:${name}` });
  }

  if (isArrowFunctionExpression(node)) {
    const name = getAnonymousName("arrow");
    return tracker.enterScope({ segment: name, kind: "function", stableKey: "arrow" });
  }

  if (isFunctionDeclaration(node) || isFunctionExpression(node)) {
    const explicitName = node.identifier?.type === "Identifier" ? node.identifier.value : undefined;
    const name = explicitName ?? getAnonymousName("function");
    return tracker.enterScope({ segment: name, kind: "function", stableKey: `func:${name}` });
  }

  if (isClassDeclaration(node)) {
    const explicitName = node.identifier?.type === "Identifier" ? node.identifier.value : undefined;
    const name = explicitName ?? getAnonymousName("class");
    return tracker.enterScope({ segment: name, kind: "class", stableKey: `class:${name}` });
  }

  // Object properties
  if (isKeyValueProperty(node)) {
    const key = node.key;
    const name = key.type === "Identifier" ? key.value : key.type === "StringLiteral" ? key.value : null;
    if (name) {
      return tracker.enterScope({ segment: name, kind: "property", stableKey: `prop:${name}` });
    }
  }

  return null;
};

// Type guards
const isCallExpression = (node: unknown): node is CallExpression =>
  typeof node === "object" && node !== null && "type" in node && node.type === "CallExpression";

const isExpression = (node: unknown): node is Expression => typeof node === "object" && node !== null && "type" in node;

const isAssignmentExpression = (node: unknown): node is AssignmentExpression =>
  typeof node === "object" && node !== null && "type" in node && node.type === "AssignmentExpression";

const isVariableDeclarator = (node: unknown): node is VariableDeclarator =>
  typeof node === "object" && node !== null && "type" in node && node.type === "VariableDeclarator";

const isArrowFunctionExpression = (node: unknown): node is ArrowFunctionExpression =>
  typeof node === "object" && node !== null && "type" in node && node.type === "ArrowFunctionExpression";

const isFunctionDeclaration = (node: unknown): node is FunctionDeclaration =>
  typeof node === "object" && node !== null && "type" in node && node.type === "FunctionDeclaration";

const isFunctionExpression = (node: unknown): node is FunctionExpression =>
  typeof node === "object" && node !== null && "type" in node && node.type === "FunctionExpression";

const isClassDeclaration = (node: unknown): node is ClassDeclaration =>
  typeof node === "object" && node !== null && "type" in node && node.type === "ClassDeclaration";

const isKeyValueProperty = (node: unknown): node is { type: "KeyValueProperty"; key: { type: "Identifier"; value: string } } =>
  typeof node === "object" &&
  node !== null &&
  "type" in node &&
  node.type === "KeyValueProperty" &&
  "key" in node &&
  typeof node.key === "object" &&
  node.key !== null &&
  "type" in node.key &&
  node.key.type === "Identifier";

File: packages/swc-transformer/src/transform/metadata.rs
//! Metadata collection module.
//!
//! This module collects metadata about GQL definitions in the source code:
//! - AST path (canonical path)
//! - Export bindings
//! - Scope tracking

use std::collections::HashMap;
use swc_core::common::Span;
use swc_core::ecma::ast::*;
use swc_core::ecma::visit::{Visit, VisitWith};

/// Metadata about a GQL definition.
#[derive(Debug, Clone)]
pub struct GqlDefinitionMetadata {
    /// The AST path for canonical ID resolution.
    pub ast_path: String,
    /// Whether this is a top-level definition.
    #[allow(dead_code)]
    pub is_top_level: bool,
    /// Whether this definition is exported.
    #[allow(dead_code)]
    pub is_exported: bool,
    /// The export binding name, if exported.
    #[allow(dead_code)]
    pub export_binding: Option<String>,
}

/// Map from call expression span to metadata.
pub type MetadataMap = HashMap<Span, GqlDefinitionMetadata>;

/// Map from local name to export name.
type ExportBindingMap = HashMap<String, String>;

/// Collects metadata about GQL definitions in a module.
pub struct MetadataCollector {
    #[allow(dead_code)]
    source_path: String,
    export_bindings: ExportBindingMap,
    scope_stack: Vec<ScopeFrame>,
    metadata: MetadataMap,
    anonymous_counters: HashMap<String, usize>,
    #[allow(dead_code)]
    definition_counter: usize,
}

struct ScopeFrame {
    segment: String,
    #[allow(dead_code)]
    kind: String,
}

impl MetadataCollector {
    /// Collect metadata from a module.
    pub fn collect(module: &Module, source_path: &str) -> MetadataMap {
        let export_bindings = Self::collect_export_bindings(module);

        let mut collector = Self {
            source_path: source_path.to_string(),
            export_bindings,
            scope_stack: Vec::new(),
            metadata: HashMap::new(),
            anonymous_counters: HashMap::new(),
            definition_counter: 0,
        };

        module.visit_with(&mut collector);
        collector.metadata
    }

    /// Collect export bindings from the module.
    fn collect_export_bindings(module: &Module) -> ExportBindingMap {
        let mut bindings = HashMap::new();

        for item in &module.body {
            match item {
                // ESM: export { foo }
                ModuleItem::ModuleDecl(ModuleDecl::ExportNamed(export)) => {
                    if export.src.is_none() {
                        for spec in &export.specifiers {
                            if let ExportSpecifier::Named(named) = spec {
                                let local = match &named.orig {
                                    ModuleExportName::Ident(id) => atom_to_string(&id.sym),
                                    ModuleExportName::Str(s) => wtf8_to_string(&s.value),
                                };
                                let exported = match &named.exported {
                                    Some(ModuleExportName::Ident(id)) => atom_to_string(&id.sym),
                                    Some(ModuleExportName::Str(s)) => wtf8_to_string(&s.value),
                                    None => local.clone(),
                                };
                                bindings.insert(local, exported);
                            }
                        }
                    }
                }

                // ESM: export const foo = ...
                ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(export)) => {
                    if let Decl::Var(var_decl) = &export.decl {
                        for decl in &var_decl.decls {
                            if let Pat::Ident(ident) = &decl.name {
                                let name = atom_to_string(&ident.id.sym);
                                bindings.insert(name.clone(), name);
                            }
                        }
                    } else if let Decl::Fn(fn_decl) = &export.decl {
                        let name = atom_to_string(&fn_decl.ident.sym);
                        bindings.insert(name.clone(), name);
                    } else if let Decl::Class(class_decl) = &export.decl {
                        let name = atom_to_string(&class_decl.ident.sym);
                        bindings.insert(name.clone(), name);
                    }
                }

                // CommonJS: exports.foo = ... or module.exports.foo = ...
                ModuleItem::Stmt(Stmt::Expr(expr_stmt)) => {
                    if let Expr::Assign(assign) = &*expr_stmt.expr {
                        if let Some(name) = get_commonjs_export_name(&assign.left) {
                            bindings.insert(name.clone(), name);
                        }
                    }
                }

                _ => {}
            }
        }

        bindings
    }

    /// Get the current AST path.
    fn get_ast_path(&self) -> String {
        self.scope_stack
            .iter()
            .map(|f| f.segment.clone())
            .collect::<Vec<_>>()
            .join(".")
    }

    /// Get an anonymous name for a scope kind (not currently used but kept for future).
    #[allow(dead_code)]
    fn get_anonymous_name(&mut self, kind: &str) -> String {
        let count = self.anonymous_counters.entry(kind.to_string()).or_insert(0);
        let name = format!("{}#{}", kind, count);
        *count += 1;
        name
    }

    /// Register a definition and get its AST path.
    /// The AST path is the scope segments joined by `.`, with `$N` suffix for duplicates.
    fn register_definition(&mut self) -> String {
        let base_path = self.get_ast_path();

        // Track occurrences for uniqueness
        let count = self.anonymous_counters.entry(base_path.clone()).or_insert(0);
        let path = if *count == 0 {
            base_path.clone()
        } else {
            format!("{}${}", base_path, count)
        };
        *count += 1;
        path
    }

    /// Enter a scope.
    fn enter_scope(&mut self, segment: String, kind: &str) {
        self.scope_stack.push(ScopeFrame {
            segment,
            kind: kind.to_string(),
        });
    }

    /// Exit a scope.
    fn exit_scope(&mut self) {
        self.scope_stack.pop();
    }

    /// Check if a call expression is a GQL definition call.
    fn is_gql_definition_call(&self, call: &CallExpr) -> bool {
        // Check if callee is gql.* pattern
        if let Callee::Expr(expr) = &call.callee {
            if let Expr::Member(member) = &**expr {
                if is_gql_reference(&member.obj) {
                    // Check if first argument is an arrow function
                    if let Some(first_arg) = call.args.first() {
                        return matches!(&*first_arg.expr, Expr::Arrow(_));
                    }
                }
            }
        }
        false
    }

    /// Resolve top-level export info for a call.
    fn resolve_export_info(&self, _call: &CallExpr) -> Option<String> {
        // This is a simplified version - in practice, you'd need to track
        // parent nodes to find the variable declaration or assignment
        // For now, we'll look at the scope stack
        if self.scope_stack.len() == 1 {
            let binding_name = &self.scope_stack[0].segment;
            self.export_bindings.get(binding_name).cloned()
        } else {
            None
        }
    }
}

impl Visit for MetadataCollector {
    fn visit_var_declarator(&mut self, decl: &VarDeclarator) {
        if let Pat::Ident(ident) = &decl.name {
            let name = atom_to_string(&ident.id.sym);
            self.enter_scope(name, "variable");
            decl.visit_children_with(self);
            self.exit_scope();
        } else {
            decl.visit_children_with(self);
        }
    }

    fn visit_fn_decl(&mut self, decl: &FnDecl) {
        let name = atom_to_string(&decl.ident.sym);
        self.enter_scope(name, "function");
        decl.visit_children_with(self);
        self.exit_scope();
    }

    fn visit_fn_expr(&mut self, expr: &FnExpr) {
        let name = expr
            .ident
            .as_ref()
            .map(|i| atom_to_string(&i.sym))
            .unwrap_or_else(|| self.get_anonymous_name("function"));
        self.enter_scope(name, "function");
        expr.visit_children_with(self);
        self.exit_scope();
    }

    fn visit_arrow_expr(&mut self, expr: &ArrowExpr) {
        let name = self.get_anonymous_name("arrow");
        self.enter_scope(name, "function");
        expr.visit_children_with(self);
        self.exit_scope();
    }

    fn visit_class_decl(&mut self, decl: &ClassDecl) {
        let name = atom_to_string(&decl.ident.sym);
        self.enter_scope(name, "class");
        decl.visit_children_with(self);
        self.exit_scope();
    }

    fn visit_class_method(&mut self, method: &ClassMethod) {
        if let PropName::Ident(ident) = &method.key {
            let name = atom_to_string(&ident.sym);
            self.enter_scope(name, "method");
            method.visit_children_with(self);
            self.exit_scope();
        } else {
            method.visit_children_with(self);
        }
    }

    fn visit_key_value_prop(&mut self, prop: &KeyValueProp) {
        let name = match &prop.key {
            PropName::Ident(ident) => Some(atom_to_string(&ident.sym)),
            PropName::Str(s) => Some(wtf8_to_string(&s.value)),
            _ => None,
        };

        if let Some(name) = name {
            self.enter_scope(name, "property");
            prop.visit_children_with(self);
            self.exit_scope();
        } else {
            prop.visit_children_with(self);
        }
    }

    fn visit_assign_expr(&mut self, expr: &AssignExpr) {
        // Handle CommonJS exports: exports.foo = ...
        if let Some(name) = get_commonjs_export_name(&expr.left) {
            self.enter_scope(name, "variable");
            expr.visit_children_with(self);
            self.exit_scope();
        } else {
            expr.visit_children_with(self);
        }
    }

    fn visit_call_expr(&mut self, call: &CallExpr) {
        if self.is_gql_definition_call(call) {
            let ast_path = self.register_definition();
            let is_top_level = self.scope_stack.len() <= 1;
            let export_binding = self.resolve_export_info(call);

            self.metadata.insert(
                call.span,
                GqlDefinitionMetadata {
                    ast_path,
                    is_top_level,
                    is_exported: export_binding.is_some(),
                    export_binding,
                },
            );

            // Don't visit children of GQL calls
            return;
        }

        call.visit_children_with(self);
    }
}

/// Check if an expression is a reference to `gql`.
fn is_gql_reference(expr: &Expr) -> bool {
    match expr {
        Expr::Ident(ident) => atom_eq(&ident.sym, "gql"),
        Expr::Member(member) => {
            if let MemberProp::Ident(ident) = &member.prop {
                if atom_eq(&ident.sym, "gql") {
                    return true;
                }
            }
            is_gql_reference(&member.obj)
        }
        _ => false,
    }
}

/// Get the export name from a CommonJS export pattern.
fn get_commonjs_export_name(target: &AssignTarget) -> Option<String> {
    match target {
        AssignTarget::Simple(SimpleAssignTarget::Member(member)) => {
            // Check for exports.foo or module.exports.foo
            let is_exports = matches!(&*member.obj, Expr::Ident(ident) if atom_eq(&ident.sym, "exports"));
            let is_module_exports = if let Expr::Member(inner) = &*member.obj {
                matches!(&*inner.obj, Expr::Ident(ident) if atom_eq(&ident.sym, "module"))
                    && matches!(&inner.prop, MemberProp::Ident(ident) if atom_eq(&ident.sym, "exports"))
            } else {
                false
            };

            if !is_exports && !is_module_exports {
                return None;
            }

            // Extract property name
            if let MemberProp::Ident(ident) = &member.prop {
                Some(atom_to_string(&ident.sym))
            } else {
                None
            }
        }
        _ => None,
    }
}

/// Helper to compare an Atom with a string.
fn atom_eq<T: AsRef<str>>(atom: &T, s: &str) -> bool {
    atom.as_ref() == s
}

/// Helper to convert an Atom to String.
fn atom_to_string<T: AsRef<str>>(atom: &T) -> String {
    atom.as_ref().to_string()
}

/// Helper to convert a Wtf8Atom (string literal value) to String.
fn wtf8_to_string(atom: &swc_core::atoms::Wtf8Atom) -> String {
    atom.to_string_lossy().into_owned()
}
